README
written by Helmut K. C. Tessarek

Last Update: 2014-05-09

http://mod-auth-ibmdb2.sourceforge.net/

Contents:
---------

1) Description of the library
2) File Description

1) Description of the library
-----------------------------

This library delivers the following functions as UDFs and SPs:

function in library          UDF / SP

bcrypt                       bcrypt
phpmd5                       php_md5
aprmd5                       apr_md5
aprcrypt                     apr_crypt
aprsha1                      apr_sha1
aprsha256                    apr_sha256
validate                     validate_pw

The phpmd5 function is compatible to the PHP md5 function.
The aprmd5, aprcrypt, aprsha1, bcrypt functions are compatible to the
functions used in Apache's htpasswd utility.
The aprsha256 routine returns the identifier {SHA256} plus the base64
encoded sha256 hash.
The validate function validates a password against a hash.

In win32 environments apr_crypt returns the output of bcrypt, if available.
If bcrypt is not available, the output of apr_md5 is returned.

2) File Description
-------------------

db2hash.c       the SQL API stuff
hash.c          the c code for the hash functions
hash.h          include file
sha2.h          include file for APR sha256
register.ddl    script to register the UDFs and SPs
drop.ddl        script to drop the UDFs and SPs
makertn         bash script to compile the library (Linux/AIX)
test_hash.c     test program for the functions
hash.exp        function export file for AIX
hash.def        definition file for Windows
makertn.bat     script to compile the library (win32)
maketest.bat    script to compile test program
INSTALL         compiling and installing instructions
CHANGES         change log
README          this file


CHANGES
db2-hash-routines 1.2

 - added bcrypt
 - added sha256 / {SHA256}base64_encoded_hash
 - renamed md5 to php_md5
 - changed man pages

db2-hash-routines 1.1

 - renamed package to db2-hash-routines
 - changed man pages

db2-auth-routines 1.1

 - renamed package to db2-auth-routines
 - added man pages
 - added stored procedures
 - renamed several files (see README)
 - changed makertn script to optionally compile with xlc on AIX

db2-auth-udfs 1.3

 - allow empty input parameters (empty strings)
 - changed reg_udfs.ddl to register the functions as NOT FENCED

db2-auth-udfs 1.2

 - changed the code to use the APR library
 - renamed the files
 - changed makeudf.bat to compile with MS compiler
 - added maketest[.bat] to compile test program

db2-auth-udfs 1.1.1

 - changed the makemod script to use it on AIX as well

db2-auth-udfs 1.1

 - added script for compiling the library on Windows
 - rewrote the README file for better understanding

db2-auth-udfs 1.0

 - initial release

http://mod-auth-ibmdb2.sourceforge.net/


test_hash.c
#include <stdio.h>
#include <string.h>
#include "hash.h"

int main( int argc, char *argv[] )
{
	char clear[121];
	char *hash;

	if (argc < 2 || strlen(argv[1]) > 120)
	{
		strcpy(clear, "test");
	}
	else
	{
		strcpy(clear, argv[1]);
	}

	printf( "\nUsing word '%s' (without the quotes)\n", clear );

	hash = mk_hash( clear, ALG_PHPMD5 );
	printf( "\nphp_md5     -> %s", hash);

	hash = mk_hash( clear, ALG_APMD5 );
	printf( "\napr_md5     -> %s", hash);

	hash = mk_hash( clear, ALG_CRYPT );
	printf( "\napr_crypt   -> %s", hash);

	hash = mk_hash( clear, ALG_APSHA );
	printf( "\napr_sha1    -> %s", hash);

	hash = mk_hash( clear, ALG_APSHA256 );
	printf( "\napr_sha256  -> %s", hash);

#if BCRYPT_ALGO_SUPPORTED
	hash = mk_hash( clear, ALG_BCRYPT );
	printf( "\nbcrypt      -> %s", hash);
#endif

	printf("\n\n");

	free(hash);

	return 0;
}


hash.h
/*
  +----------------------------------------------------------------------+
  | hash.h: hashing functions include (hashing library for IBM DB2 )     |
  +----------------------------------------------------------------------+
  | Copyright (c) 2007-2014 Helmut K. C. Tessarek                        |
  +----------------------------------------------------------------------+
  | Licensed under the Apache License, Version 2.0 (the "License"); you  |
  | may not use this file except in compliance with the License. You may |
  | obtain a copy of the License at                                      |
  | http://www.apache.org/licenses/LICENSE-2.0                           |
  |                                                                      |
  | Unless required by applicable law or agreed to in writing, software  |
  | distributed under the License is distributed on an "AS IS" BASIS,    |
  | WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or      |
  | implied. See the License for the specific language governing         |
  | permissions and limitations under the License.                       |
  +----------------------------------------------------------------------+
  | Author: Helmut K. C. Tessarek                                        |
  +----------------------------------------------------------------------+
  | Website: http://mod-auth-ibmdb2.sourceforge.net                      |
  +----------------------------------------------------------------------+
*/

#ifndef HASH_H
#define HASH_H

#include "apr.h"
#include "apr_md5.h"
#include "apr_sha1.h"
#include "apr_strings.h"
#include "apu_version.h"
#include <time.h>

#if APR_HAVE_CRYPT_H
#include <crypt.h>
#endif
#if APR_HAVE_STDLIB_H
#include <stdlib.h>
#endif
#if APR_HAVE_STRING_H
#include <string.h>
#endif
#if APR_HAVE_UNISTD_H
#include <unistd.h>
#endif

#if APU_MAJOR_VERSION == 1 && APU_MINOR_VERSION >= 5
#define BCRYPT_ALGO_SUPPORTED 1
#else
#define BCRYPT_ALGO_SUPPORTED 0
#endif

#define ALG_CRYPT    1
#define ALG_PHPMD5   2
#define ALG_APMD5    3
#define ALG_APSHA    4
#define ALG_APSHA256 5
#define ALG_BCRYPT   7

#define APR_SHA256PW_ID        "{SHA256}"
#define APR_SHA256PW_IDLEN     8

static int generate_salt(char *s, size_t size);
void sha256_base64( const char *clear, int len, char *out );
char* mk_hash( const char *passwd, int alg );

#endif


sha2.h
/*
  +----------------------------------------------------------------------+
  | sha2.h: APR sha256 include file                                      |
  +----------------------------------------------------------------------+
  | Licensed under the Apache License, Version 2.0 (the "License"); you  |
  | may not use this file except in compliance with the License. You may |
  | obtain a copy of the License at                                      |
  | http://www.apache.org/licenses/LICENSE-2.0                           |
  |                                                                      |
  | Unless required by applicable law or agreed to in writing, software  |
  | distributed under the License is distributed on an "AS IS" BASIS,    |
  | WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or      |
  | implied. See the License for the specific language governing         |
  | permissions and limitations under the License.                       |
  +----------------------------------------------------------------------+
  | Author: Aaron D. Gifford <me@aarongifford.com>                       |
  +----------------------------------------------------------------------+
  | A licence was granted to the ASF by Aaron on 4 November 2003.        |
  +----------------------------------------------------------------------+
*/

#ifndef __SHA2_H__
#define __SHA2_H__

#ifdef __cplusplus
extern "C" {
#endif

#include "apr.h"

/*** SHA-256 Various Length Definitions ***********************/
#define SHA256_BLOCK_LENGTH             64
#define SHA256_DIGEST_LENGTH            32
#define SHA256_DIGEST_STRING_LENGTH     (SHA256_DIGEST_LENGTH * 2 + 1)


/*** SHA-256/384/512 Context Structures *******************************/
typedef struct _SHA256_CTX {
        apr_uint32_t    state[8];
        apr_uint64_t    bitcount;
        apr_byte_t      buffer[SHA256_BLOCK_LENGTH];
} SHA256_CTX;


/*** SHA-256/384/512 Function Prototypes ******************************/
void apr__SHA256_Init(SHA256_CTX *);
void apr__SHA256_Update(SHA256_CTX *, const apr_byte_t *, size_t);
void apr__SHA256_Final(apr_byte_t [SHA256_DIGEST_LENGTH], SHA256_CTX *);
char* apr__SHA256_End(SHA256_CTX *, char [SHA256_DIGEST_STRING_LENGTH]);
char* apr__SHA256_Data(const apr_byte_t *, size_t,
                  char [SHA256_DIGEST_STRING_LENGTH]);

#ifdef  __cplusplus
}
#endif /* __cplusplus */

#endif /* __SHA2_H__ */

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4 fdm=marker
 * vim<600: noet sw=4 ts=4
 */


db2hash.c
/*
  +----------------------------------------------------------------------+
  | db2hash.c: hashing library for IBM DB2                               |
  +----------------------------------------------------------------------+
  | Copyright (c) 2007-2014 Helmut K. C. Tessarek                        |
  +----------------------------------------------------------------------+
  | Licensed under the Apache License, Version 2.0 (the "License"); you  |
  | may not use this file except in compliance with the License. You may |
  | obtain a copy of the License at                                      |
  | http://www.apache.org/licenses/LICENSE-2.0                           |
  |                                                                      |
  | Unless required by applicable law or agreed to in writing, software  |
  | distributed under the License is distributed on an "AS IS" BASIS,    |
  | WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or      |
  | implied. See the License for the specific language governing         |
  | permissions and limitations under the License.                       |
  +----------------------------------------------------------------------+
  | Author: Helmut K. C. Tessarek                                        |
  +----------------------------------------------------------------------+
  | Website: http://mod-auth-ibmdb2.sourceforge.net                      |
  +----------------------------------------------------------------------+
*/

#include <stdio.h>
#include <sqludf.h>
#include <sqlca.h>
#include <sqlda.h>
#include "hash.h"

/*
  +----------------------------------------------------------------------+
  | function phpmd5: MD5 hashing (as in PHP)                             |
  |                                                                      |
  |          input : varchar             (password)                      |
  |          output: char                (hash)                          |
  +----------------------------------------------------------------------+
*/

#ifdef __cplusplus
extern "C"
#endif
void SQL_API_FN phpmd5(	SQLUDF_CHAR      *in,
						SQLUDF_CHAR      out[33],
						SQLUDF_SMALLINT  *innull,
						SQLUDF_SMALLINT  *outnull,
						SQLUDF_TRAIL_ARGS)
{
	char *t;

	if( *innull != 0 )
	{
		*outnull = -1;
		return;
	}

	t = mk_hash( in, ALG_PHPMD5 );
	strcpy( out, t );
	free( t );

	*outnull = 0;
	return;
}

/*
  +----------------------------------------------------------------------+
  | function aprmd5: MD5 hashing (as in Apache's htpasswd program)       |
  |                                                                      |
  |          input : varchar             (password)                      |
  |          output: char                (hash)                          |
  +----------------------------------------------------------------------+
*/

#ifdef __cplusplus
extern "C"
#endif
void SQL_API_FN aprmd5(	SQLUDF_CHAR      *in,
						SQLUDF_CHAR      out[38],
						SQLUDF_SMALLINT  *innull,
						SQLUDF_SMALLINT  *outnull,
						SQLUDF_TRAIL_ARGS)
{
	char *t;

	if( *innull != 0 )
	{
		*outnull = -1;
		return;
	}

	t = mk_hash( in, ALG_APMD5 );
	strcpy( out, t );
	free( t );

	*outnull = 0;
	return;
}

/*
  +----------------------------------------------------------------------+
  | function aprcrypt: crypt function (as in Apache's htpasswd program)  |
  |                                                                      |
  |          input : varchar             (password)                      |
  |          output: char                (hash)                          |
  +----------------------------------------------------------------------+
*/

#ifdef __cplusplus
extern "C"
#endif
void SQL_API_FN aprcrypt(	SQLUDF_CHAR      *in,
							SQLUDF_CHAR      out[14],
							SQLUDF_SMALLINT  *innull,
							SQLUDF_SMALLINT  *outnull,
							SQLUDF_TRAIL_ARGS)
{
	char *t;

	if( *innull != 0 )
	{
		*outnull = -1;
		return;
	}

	t = mk_hash( in, ALG_CRYPT );
	strcpy( out, t );
	free( t );

	*outnull = 0;
	return;
}

/*
  +----------------------------------------------------------------------+
  | function aprsha1: SHA1 hashing (as in Apache's htpasswd program)     |
  |                                                                      |
  |          input : varchar             (password)                      |
  |          output: char                (hash)                          |
  +----------------------------------------------------------------------+
*/

#ifdef __cplusplus
extern "C"
#endif
void SQL_API_FN aprsha1(	SQLUDF_CHAR      *in,
							SQLUDF_CHAR      out[34],
							SQLUDF_SMALLINT  *innull,
							SQLUDF_SMALLINT  *outnull,
							SQLUDF_TRAIL_ARGS)
{
	char *t;

	if( *innull != 0 )
	{
		*outnull = -1;
		return;
	}

	t = mk_hash( in, ALG_APSHA );
	strcpy( out, t );
	free( t );

	*outnull = 0;
	return;
}

/*
  +----------------------------------------------------------------------+
  | function aprsha256: SHA256 hashing (similar to the algorithm used    |
  |                                     in Apache's htpasswd program)    |
  |                                                                      |
  |          input : varchar             (password)                      |
  |          output: char                (hash)                          |
  +----------------------------------------------------------------------+
*/

#ifdef __cplusplus
extern "C"
#endif
void SQL_API_FN aprsha256(	SQLUDF_CHAR      *in,
							SQLUDF_CHAR      out[53],
							SQLUDF_SMALLINT  *innull,
							SQLUDF_SMALLINT  *outnull,
							SQLUDF_TRAIL_ARGS)
{
	char *t;

	if( *innull != 0 )
	{
		*outnull = -1;
		return;
	}

	t = mk_hash( in, ALG_APSHA256 );
	strcpy( out, t );
	free( t );

	*outnull = 0;
	return;
}

/*
  +----------------------------------------------------------------------+
  | function bcrypt: bcrypt (as in Apache's htpasswd program)            |
  |                                                                      |
  |          input : varchar             (password)                      |
  |          output: char                (hash)                          |
  +----------------------------------------------------------------------+
*/

#if BCRYPT_ALGO_SUPPORTED
#ifdef __cplusplus
extern "C"
#endif
void SQL_API_FN bcrypt(	SQLUDF_CHAR      *in,
						SQLUDF_CHAR      out[61],
						SQLUDF_SMALLINT  *innull,
						SQLUDF_SMALLINT  *outnull,
						SQLUDF_TRAIL_ARGS)
{
	char *t;

	if( *innull != 0 )
	{
		*outnull = -1;
		return;
	}

	t = mk_hash( in, ALG_BCRYPT );
	strcpy( out, t );
	free( t );

	*outnull = 0;
	return;
}
#endif

/*
  +----------------------------------------------------------------------+
  | function validate: validates the hash                                |
  |                                                                      |
  |          input1: varchar             (password)                      |
  |          input2: varchar             (hash)                          |
  |          output: integer             (status)                        |
  +----------------------------------------------------------------------+
*/

#ifdef __cplusplus
extern "C"
#endif
SQL_API_RC SQL_API_FN validate(	SQLUDF_CHAR      *password,
								SQLUDF_CHAR      *hash,
								SQLUDF_INTEGER	 *out,
								SQLUDF_SMALLINT  *passwordNullInd,
								SQLUDF_SMALLINT  *hashNullInd,
								SQLUDF_SMALLINT  *outNullInd,
								SQLUDF_TRAIL_ARGS)
{
	apr_status_t status;
	char *tmphash, *phpmd5, *result;

	*out = -1;
	*outNullInd = -1;

	if( *passwordNullInd != 0 || *hashNullInd != 0 )
	{
		*outNullInd = -1;
		return(0);
	}

	if( strlen(hash) == 0 )
	{
		strcpy(SQLUDF_STATE, "39701");
		strcpy(SQLUDF_MSGTX, "The second parameter (hash) must not be empty.");
		*outNullInd = 0;
		return(0);
	}

	if( !strncmp( hash, APR_SHA256PW_ID, APR_SHA256PW_IDLEN ) )
	{
		tmphash = mk_hash( password, ALG_APSHA256 );

		if( strcmp( hash, tmphash ) == 0 )
			*out = 1;
		else
			*out = 0;

		free(tmphash);

		*outNullInd = 0;
		return(0);
	}

	if( strlen(hash) == 32 && (hash[0] != '$') )
	{
		phpmd5 = mk_hash( password, ALG_PHPMD5 );

		if( apr_strnatcmp( hash, phpmd5 ) == 0 )
		{
			*out = 1;
		}
		else
		{
			*out = 0;
		}

		free(phpmd5);

		*outNullInd = 0;
		return(0);
	}

	status = apr_password_validate( password, hash );

	if( status == APR_SUCCESS )
	{
		*out = 1;
	}
#ifndef WIN32
	else
	{
		// maybe a different encrypted password (glibc2 crypt)?
		result = crypt( password, hash );
		if( strcmp( hash, result ) == 0 )
		{
			*out = 1;
		}
		else
		{
			*out = 0;
		}
	}
#endif

	*outNullInd = 0;
	return(0);
}


hash.c
/*
  +----------------------------------------------------------------------+
  | hash.c: hashing functions (hashing library for IBM DB2)              |
  +----------------------------------------------------------------------+
  | Copyright (c) 2007-2014 Helmut K. C. Tessarek                        |
  +----------------------------------------------------------------------+
  | Licensed under the Apache License, Version 2.0 (the "License"); you  |
  | may not use this file except in compliance with the License. You may |
  | obtain a copy of the License at                                      |
  | http://www.apache.org/licenses/LICENSE-2.0                           |
  |                                                                      |
  | Unless required by applicable law or agreed to in writing, software  |
  | distributed under the License is distributed on an "AS IS" BASIS,    |
  | WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or      |
  | implied. See the License for the specific language governing         |
  | permissions and limitations under the License.                       |
  +----------------------------------------------------------------------+
  | Author: Helmut K. C. Tessarek                                        |
  +----------------------------------------------------------------------+
  | Website: http://mod-auth-ibmdb2.sourceforge.net                      |
  +----------------------------------------------------------------------+
*/

#include <stdio.h>
#include <string.h>
#include "hash.h"
#include "sha2.h"

static int generate_salt(char *s, size_t size)
{
	unsigned char rnd[32];
	static const char itoa64[] =
		"./0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
	apr_size_t n;
	unsigned int val = 0, bits = 0;
	apr_status_t rv;

	n = (size * 6 + 7)/8;
	if (n > sizeof(rnd))
	{
		return -1;
	}
	rv = apr_generate_random_bytes(rnd, n);
	if (rv)
	{
		return -1;
	}
	n = 0;
	while (size > 0)
	{
		if (bits < 6)
		{
			val |= (rnd[n++] << bits);
			bits += 8;
		}
		*s++ = itoa64[val & 0x3f];
		size--;
		val >>= 6;
		bits -= 6;
	}
	*s = '\0';
	return 0;
}

void sha256_base64( const char *clear, int len, char *out )
{
	int l;
	SHA256_CTX context;
	apr_byte_t digest[SHA256_DIGEST_LENGTH];

	apr__SHA256_Init( &context );
	apr__SHA256_Update( &context, clear, len );
	apr__SHA256_Final( digest, &context );

	apr_cpystrn( out, APR_SHA256PW_ID, APR_SHA256PW_IDLEN + 1 );

	l = apr_base64_encode_binary( out + APR_SHA256PW_IDLEN, digest, sizeof(digest) );
	out[l + APR_SHA256PW_IDLEN] = '\0';
}

char* mk_hash( const char *passwd, int alg )
{
	char *result;
	char cpw[120];
	char salt[16];
	int ret = 0;

	int cost = 5;

	unsigned char digest[APR_MD5_DIGESTSIZE];
	apr_md5_ctx_t context;
	char md5str[33];
	int i;
	char *r;

	cpw[0] = '\0';

	switch (alg)
	{
#if BCRYPT_ALGO_SUPPORTED
		case ALG_BCRYPT:
		default:
			ret = apr_generate_random_bytes((unsigned char*)salt, 16);
			if (ret != 0)
				break;

			ret = apr_bcrypt_encode(passwd, cost, (unsigned char*)salt, 16,
								   cpw, sizeof(cpw));
			if (ret != 0)
				break;

			break;
#endif

		case ALG_APSHA256:
			sha256_base64(passwd, strlen(passwd), cpw);
			break;

		case ALG_APSHA:
			apr_sha1_base64(passwd, strlen(passwd), cpw);
			break;

		case ALG_APMD5:
#if !BCRYPT_ALGO_SUPPORTED
		default:
#endif
			ret = generate_salt(salt, 8);
			if (ret != 0)
				break;

			apr_md5_encode((const char *)passwd, (const char *)salt, cpw, sizeof(cpw));
			break;

#if !(defined(WIN32) || defined(NETWARE))
		case ALG_CRYPT:
			ret = generate_salt(salt, 8);
			if (ret != 0)
				break;

			apr_cpystrn(cpw, (char *)crypt(passwd, salt), sizeof(cpw) - 1);
			break;
#endif

		case ALG_PHPMD5:
			md5str[0] = '\0';

			apr_md5_init( &context );
			apr_md5_update( &context, passwd, strlen(passwd) );
			apr_md5_final( digest, &context );
			for( i = 0, r = md5str; i < APR_MD5_DIGESTSIZE; i++, r += 2 )
			{
			sprintf( r, "%02x", digest[i] );
			}
			*r = '\0';

			apr_cpystrn(cpw, md5str, sizeof(md5str));
			break;
	}

	result = (char*)malloc(strlen(cpw)*sizeof(char));
	strcpy( result, cpw );
	memset(cpw, '\0', strlen(cpw));

	return result;
}
